src/cooler/util.py

```diff
@@ -1,850 +1,871 @@
+    Genomic regions are represented as half-open intervals (0-based starts,
+    1-based ends) along the length coordinate of a contig/scaffold/chromosome.
+
+    Parameters
+    ----------
+    reg : str or tuple
+        UCSC-style genomic region string, or
+        Triple (chrom, start, end), where ``start`` or ``end`` may be ``None``.
+    chromsizes : mapping, optional
+        Lookup table of scaffold lengths to check against ``chrom`` and the
+        ``end`` coordinate. Required if ``end`` is not supplied.
+
+    Returns
+    -------
+    A well-formed genomic region triple (str, int, int)
+
+    """
+    if isinstance(reg, str):
+        chrom, start, end = parse_region_string(reg)
+    else:
+        chrom, start, end = reg
+        start = int(start) if start is not None else start
+        end = int(end) if end is not None else end
+
+    try:
+        clen = chromsizes[chrom] if chromsizes is not None else None
+    except KeyError as e:
+        raise ValueError(f"Unknown sequence label: {chrom}") from e
+
+    start = 0 if start is None else start
+    if end is None:
+        if clen is None:  # TODO --- remove?
+            raise ValueError("Cannot determine end coordinate.")
+        end = clen
+
+    if end < start:
+        raise ValueError("End cannot be less than start")
+
+    if start < 0 or (clen is not None and end > clen):
+        raise ValueError(f"Genomic region out of bounds: [{start}, {end})")
+
+    return chrom, start, end
+
+
+def natsort_key(s: str, _NS_REGEX=re.compile(r"(\d+)", re.U)) -> tuple:
+    return tuple([int(x) if x.isdigit() else x for x in _NS_REGEX.split(s) if x])
+
+
+def natsorted(iterable: Iterable[str]) -> list[str]:
+    return sorted(iterable, key=natsort_key)
+
+
+def argnatsort(array: Iterable[str]) -> np.ndarray:
+    array = np.asarray(array)
+    if not len(array):
+        return np.array([], dtype=int)
+    cols = tuple(zip(*(natsort_key(x) for x in array)))
+    return np.lexsort(cols[::-1])
+
+def read_chromsizes(
+    filepath_or: str | io.StringIO,
+    name_patterns: tuple[str, ...] = (r"^chr[0-9]+$", r"^chr[XY]$", r"^chrM$"),
+    all_names: bool = False,
+    verbose: bool = False,  # Optional parameter to enable verbose output
+    **kwargs,
+) -> pd.Series:
+    """
+    Parse a `<db>.chrom.sizes` or `<db>.chromInfo.txt` file from the UCSC
+    database, where `db` is a genome assembly name.
+
+    Parameters
+    ----------
+    filepath_or : str or file-like
+        Path or url to text file, or buffer.
+    name_patterns : sequence, optional
+        Sequence of regular expressions to capture desired sequence names.
+    all_names : bool, optional
+        Whether to return all contigs listed in the file.
+    verbose : bool, optional
+        Whether to enable verbose logging for diagnostics.
+    """
+    # Check if the input is a file-like object (StringIO or file path) and inspect the first line for delimiters
+    if isinstance(filepath_or, (str, io.StringIO)):

```

Comment:
A couple issues:
* StringIO is not the only kind of file-like object that can be given to pandas, so this is too restrictive.
* The `str` case does not account for URLs which can also be given to pandas.

Reply:
Perhaps look into `pd.read_csv(filepath_or, sep="\t", nrows=1, header=None)` as a way to test correctness.